src/array/vec_backed.rs

```diff
@@ -52,12 +52,24 @@ impl<T: Send + Sync + Clone + 'static> Array for VecBackedArray<T> {
         !self.is_null(i)
     }
 
-    fn slice(&self, offset: usize, length: usize) -> Box<dyn Array> {
+    fn slice(&mut self, offset: usize, length: usize) {
+        assert!(
+            offset + length <= self.len(),
+            "offset + length may not exceed length of array"
+        );
+        unsafe { self.slice_unchecked(offset, length) }
+    }
+
+    unsafe fn slice_unchecked(&mut self, offset: usize, length: usize) {
+        self.values.drain(offset..(offset + length));
+    }
```

Comment:
cc @jaychia @xcharleslin for the implementation here! Arrow2 changed `slice()` and `slice_unchecked()` to mutate the array in-place. I think that draining the underlying vector on the provided range is what we'd want to do, right?